test/unit/TestParser.hs

```diff
@@ -0,0 +1,254 @@
+            "def 'a = 0 end"
+            ( T.unlines
+                [ "1:5:"
+                , "  |"
+                , "1 | def 'a = 0 end"
+                , "  |     ^"
+                , "unexpected '''"
+                , "expecting variable name"
+                ]
+            )
+        )
+    , testCase
+        "Parse pair syntax #225"
+        (valid "def f : (int -> bool) * (int -> bool) = (\\x. false, \\x. true) end")
+    , testCase
+        "Nested pair syntax"
+        (valid "(1,2,3,4)")
+    , testCase
+        "Binder at end of block"
+        (valid "r <- build {move}")
+    , testGroup
+        "failure location - #268"
+        [ testCase
+            "located type error"
+            ( process
+                "def a =\n 42 + \"oops\"\nend"
+                "2: Can't unify int and string"
+            )
+        , testCase
+            "failure inside bind chain"
+            ( process
+                "move;\n1;\nmove"
+                "2: Can't unify int and cmd"
+            )
+        , testCase
+            "failure inside function call"
+            ( process
+                "if true \n{} \n(move)"
+                "3: Can't unify {u0} and cmd ()"
+            )
+        , testCase
+            "parsing operators #236 - report failure on invalid operator start"
+            ( process
+                "1 <== 2"
+                ( T.unlines
+                    [ "1:3:"
+                    , "  |"
+                    , "1 | 1 <== 2"
+                    , "  |   ^"
+                    , "unexpected '<'"
+                    ]
+                )
+            )
+        ]
+    , testGroup
+        "require - #201"
+        [ testCase
+            "require device"
+            (valid "require \"boat\"")
+        , testCase
+            "require entities"
+            (valid "require 64 \"rock\"")
+        , testCase
+            "invalid syntax to require"
+            ( process
+                "require x"
+                ( T.unlines
+                    [ "1:9:"
+                    , "  |"
+                    , "1 | require x"
+                    , "  |         ^"
+                    , "unexpected 'x'"
+                    , "expecting device name in double quotes or integer literal"
+                    ]
+                )
+            )
+        , testCase
+            "invalid syntax to require n"
+            ( process
+                "require 2 x"
+                ( T.unlines
+                    [ "1:11:"
+                    , "  |"
+                    , "1 | require 2 x"
+                    , "  |           ^"
+                    , "unexpected 'x'"
+                    , "expecting entity name in double quotes"
+                    ]
+                )
+            )
+        ]
+    , testGroup
+        "json encoding"
+        [ testCase "simple expr" (roundTrip "42 + 43")
+        , testCase "module def" (roundTrip "def x = 41 end; def y = 42 end")
+        ]
+    , testGroup
```

Comment:
Not your fault, but I just realized that this test group isn't about the parser at all, but rather about type checking.  But it's nice that the same framework can deal with both.  I don't know if its worth moving them, or renaming the module?

Reply:
The whole test tree is named "Language - pipeline", what should I name the module and the exported test tree? :thinking: 

Ideas:
- `TestLanguagePipeline`
- `TestLanguage`
- `TestTypecheck`

Reply:
Ah, sure, any of those would be good.  Personally I'd lean towards `TestLanguagePipeline`, although it's kinda long.

Reply:
`TestLanguagePipeline` it is! :heavy_check_mark: